### ISSUE TITLE
refactor run()

### ISSUE THREAD

**Issue by Michael-F-Bryan** [joek13/hastebin-client#2]
The `run()` function in your `main.rs` has a lot of rightward drift and could do with a bit of refactoring to make error handling better and make it easier to extend later on.

This probably won't compile as-is, but it's roughly how I've structured command-line apps in the past when using `clap`.

```rust
fn run(matches: ArgMatches) -> Result<(), String> {
  match matches.subcommand() {
    ("upload", Some(args)) => upload(args),
    _ => Err("Invalid subcommand".to_string()),
  }
}

fn upload(args: &ArgMatches) -> Result>(), String> {
  let file = args.value_of("file")
    .expect("Should never fail because `file` is required");

  let id = hastebin::upload_file(file)?;
  let url = format!("https://hastebin.com/{}", id);
  println!("{}", url);

  if args.is_present("open") {
    Command::new("xdg-open")
      .arg(url)
      .spawn()?;
  }
  
  Ok(())             
}
```

You should also probably think about returning a proper error type, having "stringly-typed" errors is a bit of an anti-pattern. I personally am a massive fan of [`error-chain`]. For more info, check out brson's amazing article on the topic, [Starting a new Rust project right, with error-chain][ec]. He's designed it so that your error type integrates seamlessly with the language and everything will *Just Work*.

[ec]: http://brson.github.io/2016/11/30/starting-with-error-chain
[`error-chain`]: https://github.com/brson/error-chain

**Comment by Michael-F-Bryan** [comment]
I only just read your `hastebin.rs` and it looks like you're already using a proper error type, so ignore the stuff I said about returning a `Result<_, String>` :) 

You could probably just keep passing the error up instead of prematurely converting it to a string though. Anything implementing `Error` must also implement `Display` so you can do `writeln!(stderr, "{}", e)`.

**Comment by joek13** [comment]
I agree with refactoring `run` to be a little bit nicer. Ill get to work on that.

~~Right now, the error type is String because it is immediately printed to the user. There is a slightly nicer `UploadError` type within the `hastebin.rs` file itself. Is it bad practice to stringify the error there, or better to do it once it's about to be printed?~~

Well, just saw your next comment. Guess we're agreed there. I might move the "stringification" down into the `write!` invocation.

**Comment by joek13** [comment]
Only potential issue is that another possible error is while running `xdg-open`. That's not an upload error. I could either create a RuntimeError enum with variants for upload errors and running errors, but it might be easier to just convert to a string.
Thoughts?

**Comment by joek13** [comment]
Closed with 801db2102f78d35e11ab628a9e9a3d8924d23f26

**Comment by Michael-F-Bryan** [comment]
What about pulling `UploadError` out into its own `errors.rs` module and renaming it to `Error`? Quite a few projects I've seen will define their own `Error` and `Result` types which represent all possible errors you can encounter in their crate.

```rust
pub mod errors {
  pub type Result<T> = Result<T, Error>;
  
  pub enum Error {
    ...  
  }
}
```

It looks like the error from `Command::spawn()` is a `std::io::Error` and you've already got an `IOError` variant, so you shouldn't need to add any new variants to your error type.
